numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
         return NPY_SUCCEED;
     }
 
-    len = PySequence_Size(obj);
-    if (len == -1) {
-        /* Check to see if it is an integer number */
-        if (PyNumber_Check(obj)) {
-            /*
-             * After the deprecation the PyNumber_Check could be replaced
-             * by PyIndex_Check.
-             * FIXME 1.9 ?
-             */
-            len = 1;
+    PyObject *seq_obj = NULL;
+
+    /*
+    * If obj is a scalar we skip all the useless computations and jump to
+    * intp_from_scalar as soon as possible.
+    */
+    if (!PyLong_CheckExact(obj) && PySequence_Check(obj)) {
+        seq_obj = PySequence_Fast(obj,
+               "Expected an integer or an iterable/sequence of integers.");
+        if (seq_obj == NULL) {
+            // we ignore this error since it is managed below
+            PyErr_Clear();
         }
     }
-    if (len < 0) {
-        PyErr_SetString(PyExc_TypeError,
-                "expected sequence object with len >= 0 or a single integer");
-        return NPY_FAIL;
-    }
-    if (len > NPY_MAXDIMS) {
-        PyErr_Format(PyExc_ValueError, "maximum supported dimension for an ndarray is %d"
-                     ", found %d", NPY_MAXDIMS, len);
-        return NPY_FAIL;
-    }
-    if (len > 0) {
-        seq->ptr = npy_alloc_cache_dim(len);
+
+    if (seq_obj == NULL) {
+        /*
+        * obj *might* be a scalar (if intp_from_scalar does not fail, at the
+        * moment no check have been performed to verify this hypothesis).
+        */
+
+        seq->ptr = npy_alloc_cache_dim(1);
         if (seq->ptr == NULL) {
             PyErr_NoMemory();
             return NPY_FAIL;
         }
+        seq->len = 1;
+
+        seq->ptr[0] = intp_from_scalar(obj);
+        /*
+        * Check whether the value set by intp_from_scalar is different than
+        * the value returned, which means that an error occurred.
+        */
+        if (PyErr_Occurred() != NULL) {
+            npy_free_cache_dim_obj(*seq);
+            seq->ptr = NULL;
+            return NPY_FAIL;
+        }
+    } else {
+        /*
+        * obj has been recognized as an iterable of integer values
+        */
+
```

Comment:
```suggestion
```
(and * indentation above)

Reply:
Fixed in e8bd7532fe0c63d89d81a502b2c1c6ecc90b737e